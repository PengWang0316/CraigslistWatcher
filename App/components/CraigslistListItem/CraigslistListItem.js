import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { ListItem, Icon } from 'react-native-elements';
import { connect } from 'react-redux';

import styles from './Styles';
import { fetchCraigslistItem } from '../../actions/CraigslistStatesActions';

/** ListItem for CraigsList's FlatList */
export class CraigslistListItem extends PureComponent {
  static propTypes = {
    item: PropTypes.object.isRequired,
    navigation: PropTypes.object.isRequired,
    fetchCraigslistItem: PropTypes.func.isRequired,
    craigslistStates: PropTypes.object.isRequired
  };

  /**
   * Update avatar and leftIcon state after a new item was added to Redux's craigslistStates.items.
   * @param {object} nextProps is the changed props values.
   * @param {object} prevState is the old state values.
   * @return {object} Return a object that contains state that need to be updated.
   */
  static getDerivedStateFromProps(nextProps, prevState) {
    if (prevState.leftIcon !== null && nextProps.craigslistStates.items[nextProps.item.dataId])
      return {
        leftIcon: null,
        avatar: { uri: nextProps.craigslistStates.items[nextProps.item.dataId].images[0] }
      };
    return null;
  }

  /**
   * Initialing states for the component.
   * @param {object} props contains props values.
   * @return {null} No return.
   */
  constructor(props) {
    super(props);
    this.state = {
      avatar: this.props.craigslistStates.items[this.props.item.dataId] ?
        { uri: this.props.craigslistStates.items[this.props.item.dataId].images[0] } : null,
      leftIcon: this.props.craigslistStates.items[this.props.item.dataId] ? null : <Icon name="wallpaper" raised onPress={this.handleLeftIconClick} />
    };
  }

  /**
   * When a user click the left icon, the avatar state will be set to the image's uri or call the action to fetch the item's detailed information.
   * @return {null} No return.
   */
  handleLeftIconClick = () => {
    if (this.props.craigslistStates.items[this.props.item.dataId]) this.setState({
      avatar: { uri: this.props.craigslistStates.items[this.props.item.dataId].images[0] },
      leftIcon: null
    }); else this.props.fetchCraigslistItem(this.props.item.url).catch(err => console.log(err));
  };

  /**
   * When a user click the component, navigate to the detailed screen.
   * @return {null} No return.
   */
  handlePress = () => this.props.navigation.navigate('DetailedItem', { dataId: this.props.item.dataId });

  /**
   * Rendering the jsx for the component.
   * @param {object} props contians all prop's vaule.
   * @return {jsx} Return jsx for the component.
   */
  render() {
    const { item } = this.props;
    return (
      <ListItem
        roundAvatar
        avatar={this.state.avatar}
        title={item.title}
        subtitle={`${item.region} ${item.datetime}`}
        leftIcon={this.state.leftIcon}
        badge={{ value: item.price }}
        containerStyle={styles.containerStyle}
        onPress={this.handlePress}
      />
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  craigslistStates: state.craigslistStates
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchCraigslistItem: url => dispatch(fetchCraigslistItem(url))
});
export default connect(mapStateToProps, mapDispatchToProps)(CraigslistListItem);
