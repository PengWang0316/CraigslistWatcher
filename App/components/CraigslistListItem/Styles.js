import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  containerStyle: {
    backgroundColor: '#FFF',
    borderBottomWidth: 0
  }
});
