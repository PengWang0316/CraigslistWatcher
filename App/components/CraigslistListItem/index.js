import CraigslistListItem from './CraigslistListItem';

export default CraigslistListItem;
