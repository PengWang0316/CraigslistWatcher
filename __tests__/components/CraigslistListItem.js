import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { CraigslistListItem } from '../../App/components/CraigslistListItem/CraigslistListItem';

jest.mock('react-native-elements', () => ({ ListItem: 'ListItem', Icon: 'Icon' }));

describe('CraigslistListItem test', () => {
  const defaultProps = {
    item: {
      datetime: '2018-04-20 17:18',
      url: 'https://sfbay.craigslist.org/eby/bab/d/sit-and-stand-double-stroller/6556745475.html',
      dataId: '6556745475',
      title: 'Sit and stand double stroller',
      price: '$45',
      region: 'albany / el cerrito'
    },
    navigation: { navigate: jest.fn() },
    fetchCraigslistItem: jest.fn().mockReturnValue(new Promise(resolve => resolve())),
    craigslistStates: {
      items: {}
    }
  };
  const getShallowComponent = (props = defaultProps) => shallow(<CraigslistListItem {...props} />);

  test('Initial component without item match', () => {
    const component = getShallowComponent();
    expect(component.state('avatar')).toBeNull();
    expect(component.state('leftIcon')).not.toBeUndefined();
    expect(component.state('leftIcon')).not.toBeNull();
  });

  test('Initial component with item match', () => {
    const component = getShallowComponent({
      ...defaultProps,
      craigslistStates: {
        items: { 6556745475: { images: ['image uri 0'] } }
      }
    });
    expect(component.state('avatar')).toEqual({ uri: 'image uri 0' });
    expect(component.state('leftIcon')).toBeNull();
  });

  test('getDerivedStateFromPros with leftIcon and match items', () => {
    // const component = getShallowComponent();
    const nextProps = {
      craigslistStates: {
        items: { 6556745475: { images: ['image uri 1'] } }
      },
      item: { dataId: 6556745475 }
    };
    const prevState = {
      leftIcon: {}
    };
    expect(CraigslistListItem.getDerivedStateFromProps(nextProps, prevState)).toEqual({
      leftIcon: null,
      avatar: { uri: 'image uri 1' }
    });
  });

  test('getDerivedStateFromPros without leftIcon and match items', () => {
    // const component = getShallowComponent();
    const nextProps = {
      craigslistStates: {
        items: { 1234: { images: ['image uri 0'] } }
      },
      item: { dataId: 6556745475 }
    };
    let prevState = {
      leftIcon: null
    };
    expect(CraigslistListItem.getDerivedStateFromProps(nextProps, prevState)).toBeNull();

    prevState = {
      leftIcon: {}
    };
    expect(CraigslistListItem.getDerivedStateFromProps(nextProps, prevState)).toBeNull();
  });

  test('handleLeftIconClick with item match', () => {
    const component = getShallowComponent({
      ...defaultProps,
      craigslistStates: {
        items: { 6556745475: { images: ['image uri 3'] } }
      }
    });
    component.setState({ leftIcon: {} });
    component.instance().handleLeftIconClick();
    expect(component.state('avatar')).toEqual({ uri: 'image uri 3' });
    expect(component.state('leftIcon')).toBeNull();
    expect(defaultProps.fetchCraigslistItem).not.toHaveBeenCalled();
  });

  test('handleLeftIconClick without item match', () => {
    const component = getShallowComponent({
      ...defaultProps,
      craigslistStates: {
        items: { 1234: { images: ['image uri 3'] } }
      }
    });
    component.instance().handleLeftIconClick();
    expect(defaultProps.fetchCraigslistItem).toHaveBeenLastCalledWith('https://sfbay.craigslist.org/eby/bab/d/sit-and-stand-double-stroller/6556745475.html');
  });

  test('handlePress', () => {
    const component = getShallowComponent();
    component.instance().handlePress();
    expect(defaultProps.navigation.navigate).toHaveBeenLastCalledWith('DetailedItem');
  });

  test(' snapshot', () => expect(renderer.create(<CraigslistListItem {...defaultProps} />).toJSON()).toMatchSnapshot());
});
