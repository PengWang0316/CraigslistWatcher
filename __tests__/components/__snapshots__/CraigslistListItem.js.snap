// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CraigslistListItem test  snapshot 1`] = `
<ListItem
  avatar={null}
  badge={
    Object {
      "value": "$45",
    }
  }
  containerStyle={
    Object {
      "backgroundColor": "#FFF",
      "borderBottomWidth": 0,
    }
  }
  leftIcon={
    <Icon
      name="wallpaper"
      onPress={[Function]}
      raised={true}
    />
  }
  onPress={[Function]}
  roundAvatar={true}
  subtitle="albany / el cerrito 2018-04-20 17:18"
  title="Sit and stand double stroller"
/>
`;
